pebble_merger.go

```diff
@@ -1,58 +1,126 @@
 package dhstore
 
 import (
+	"bytes"
 	"io"
 
 	"github.com/cockroachdb/pebble"
 )
 
-const deletableDefaultMergerName = "dhstore.v1.deletable_default_merger"
+const valueKeysMergerName = "dhstore.v1.valueKeysMerger"
 
 var (
-	_ pebble.ValueMerger          = (*deletableDefaultMerger)(nil)
-	_ pebble.DeletableValueMerger = (*deletableDefaultMerger)(nil)
+	_ pebble.ValueMerger          = (*valueKeysValueMerger)(nil)
+	_ pebble.DeletableValueMerger = (*valueKeysValueMerger)(nil)
 )
 
-type deletableDefaultMerger struct {
-	def pebble.ValueMerger
+type valueKeysValueMerger struct {
+	merges             []EncryptedValueKey
+	reverse            bool
+	marshalledSizeHint int // Used as a hint to grow the buffer size during marshalling.
+	s                  *PebbleDHStore
 }
 
-func newDeletableDefaultMerger() *pebble.Merger {
+func (s *PebbleDHStore) newValueKeysMerger() *pebble.Merger {
 	return &pebble.Merger{
 		Merge: func(k, value []byte) (pebble.ValueMerger, error) {
 			// Fall back on default merger if the key is not of type multihash, i.e. the only key
 			// type that corresponds to value-keys.
 			switch keyPrefix(k[0]) {
 			case multihashKeyPrefix:
-				merge, err := pebble.DefaultMerger.Merge(k, value)
-				if err != nil {
-					return nil, err
-				}
-				v := &deletableDefaultMerger{
-					def: merge,
-				}
-				return v, nil
+				v := &valueKeysValueMerger{s: s}
+				return v, v.MergeNewer(value)
 			default:
 				return pebble.DefaultMerger.Merge(k, value)
 			}
 		},
-		Name: deletableDefaultMergerName,
+		Name: valueKeysMergerName,
 	}
 }
 
-func (v *deletableDefaultMerger) MergeNewer(value []byte) error {
-	return v.def.MergeNewer(value)
+func (v *valueKeysValueMerger) MergeNewer(value []byte) error {
+	if len(value) == 0 {
+		return nil
+	}
+
+	evks, err := v.s.unmarshalEncryptedIndexKeys(value)
+	if err != nil {
+		return err
+	}
+
+	v.merges = maybeGrow(v.merges, len(evks))
+
+	if len(v.merges) == 0 {
+		// Optimise for the case where there are no merges.
+		v.merges = append(v.merges, evks...)
+		v.marshalledSizeHint += len(value)
+	} else {
+		for _, evk := range evks {
+			if !v.exists(evk) {
```

Comment:
How many pending merges are there expected to be on average? Write-heavy records will have very bad complexity here.

Reply:
We either keep a map which effectively doubles the memory footprint of merger or loop over and search. So far storetheindnex without the length check has beep performant.